### PR TITLE
POC: Self-contained headers

### PR DESCRIPTION
**Pull request by hebasto**
This PR adds a tool that verifies every header whether it is self-contained.

As an example, the `field_5x52.h` and `field_10x26.h` headers have been refactored to get self-contained.

### CODE REVIEW

**Comment by real-or-random** [comment]
Interesting! 

This raises the question: Should self-containedness be tested in VERIFY mode, or non-VERIFY mode, or both?  I believe at least in VERIFY mode because the goal here will be to help development tooling (e.g., clangd), and development is typically more convenient in VERIFY mode. Otherwise all `#ifdef VERIFY` blocks don't exist, i.e., you don't even get syntax highlighting for them...

Of course, we could manage to make this happen in both modes without too much hassle, that will be ideal.    


This is related to #1309.

**Comment by hebasto** [comment]
> This is related to #1039, where I suggested something like this.

I've been working on this branch a few months but decided to PR it now as I feel it might help with https://github.com/bitcoin-core/secp256k1/issues/1421#issuecomment-1714079714.

File: src/field_10x26.h [hunk 1]
```diff
@@ -30,7 +30,10 @@ typedef struct {
      *     sum(i=0..9, n[i] << (i*26)) < p
      *     (together these imply n[9] <= 2^22 - 1)
      */
-    SECP256K1_FE_VERIFY_FIELDS
+#ifdef VERIFY
+    int magnitude;
+    int normalized;
+#endif
 } secp256k1_fe;
```
Comment (real-or-random) [inline src/field_10x26.h:37]: Have you tried/considered including `field.h` instead?

Comment (hebasto) [reply src/field_10x26.h:37]: Doesn't it create a circular dependency?

Comment (real-or-random) [reply src/field_10x26.h:37]: In some sense, yes, but it's not a problem because the include guards should break any loops.

If you take that to the extreme, we could also have a single `headers.h` that includes *all* internal headers and include `headers.h` (almost) everywhere. That's a rather simple solution. I'm not convinced that this is very elegant, but it gets the job done...

Comment (hebasto) [reply src/field_10x26.h:37]: > it's not a problem because the include guards should break any loops.

Right. But not for renaming a header to a source file. It is not clear (at least for me) how to make it meaningful. FWIW, I've already tried the testing script that keeps an original header unmodified.

Comment (real-or-random) [reply src/field_10x26.h:37]: > Right. But not for renaming a header to a source file. It is not clear (at least for me) how to make it meaningful.

You mean problems will occur if we move code from a header to a source file?

> FWIW, I've already tried the testing script that keeps an original header unmodified.

I don't understand. Can you rephrase?

Comment (hebasto) [reply src/field_10x26.h:37]: > > Right. But not for renaming a header to a source file. It is not clear (at least for me) how to make it meaningful.
> 
> You mean problems will occur if we move code from a header to a source file?

Yes. I mean:
```sh
mv src/field_10x26.h src/field_10x26.c
gcc -c src/field_10x26.c -o src/field_10x26.o
```

> > FWIW, I've already tried the testing script that keeps an original header unmodified.
> 
> I don't understand. Can you rephrase?

A modified script looks like that:
```sh
cp src/field_10x26.h src/field_10x26.c
gcc -c src/field_10x26.c -o src/field_10x26.o
```

**Comment by hebasto** [comment]
Rebased.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -738,6 +738,18 @@ jobs:
             g++ -Werror include/*.h
             clang -Werror -x c++-header include/*.h
 
+  headers:
+    name: "Self-contained headers"
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
```
Comment (fanquake) [inline .github/workflows/ci.yml]:
```suggestion
        uses: actions/checkout@v5
```

Comment (hebasto) [reply .github/workflows/ci.yml]: Thanks!. Fixed.

**Comment by hebasto** [comment]
Addressed [feedback](https://github.com/bitcoin-core/secp256k1/pull/1423#discussion_r2348405846) from @fanquake.

Additionally, the check script now compiles in a temporary directory rather than in the source tree.
